### ISSUE TITLE
Axes settings seems not loaded by django in production

### ISSUE THREAD

**Issue by jeremyrich** [jazzband/django-axes#634]
Axes is in INSTALLED_APPS and AUTHENTICATION_BACKENDS.
Version : 5.5.0

When trying to run my CI and just test runner I got this :

Here is the traceback : 
Traceback (most recent call last):
  File "manage.py", line 38, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/usr/local/lib/python3.7/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.7/site-packages/django/apps/registry.py", line 122, in populate
    app_config.ready()
  File "/var/www/jung/badoom/core/apps.py", line 20, in ready
    autodiscover_modules("signals")
  File "/usr/local/lib/python3.7/site-packages/django/utils/module_loading.py", line 47, in autodiscover_modules
    import_module('%s.%s' % (app_config.name, module_to_search))
  File "/usr/local/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.7/site-packages/axes/signals.py", line 15, in <module>
    from axes.handlers.proxy import AxesProxyHandler
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/axes/handlers/proxy.py", line 16, in <module>
    log = getLogger(settings.AXES_LOGGER)
  File "/usr/local/lib/python3.7/site-packages/django/conf/__init__.py", line 80, in __getattr__
    val = getattr(self._wrapped, name)
AttributeError: 'Settings' object has no attribute 'AXES_LOGGER'

I know we can skip Axes in test but I would prefer to fix this issue, any insight ?

**Comment by aleksihakli** [comment]
What is your

1. Python version and distribution,
2. Django version, and
3. CI test runner?

Does this problem exist in your local development setup?

I'm noticing that you are running something under `/var/www/jung/badoom`, how have you configured your project path and settings?

**Comment by jeremyrich** [comment]
> What is your
> 
>     1. Python version and distribution,
> 
>     2. Django version, and
> 
>     3. CI test runner?
> 
> 
> Does this problem exist in your local development environment as well?
> 
> I'm noticing that you are running something under `/var/www/jung/badoom`, how have you configured your project path and settings?

Python : Python 3.7.6 / python:3.7-slim-buster
Django : 2.2.13 
CI test runner : My travis job that runs all django/python unittests -- but I do have the same locally when doing ./manage.py test ...

**Comment by jeremyrich** [comment]
@aleksihakli  
Here is an insight of what could be the conflict :

we recently added that to our settings

`MIGRATION_MODULES = {AppConfig.create(entry).label: None for entry in INSTALLED_APPS}`

any idea of why axes doesn't like it ?

**Comment by aleksihakli** [comment]
Well that narrows some things down a bit. I am not familiar with the migration module configuration so I don't have an idea why it would conflict with settings. Is it possible that you could toggle the MIGRATION_MODULES on and off by e.g. commenting it and see if that has an impact?

**Comment by jeremyrich** [comment]
Well if your suggestion is to be sûre it impacts axes, yes ! We removed this part of the code to test and axes stopped complaining. But we need this migration processus. 
It might be related to cyclic import / conflict with Django and axes settings ... Any insight / suggestion to debug that would be helpfull.

**Comment by aleksihakli** [comment]
Well the current way of handling settings is via django-appconf

https://github.com/django-compressor/django-appconf

So if there is a problem with loading the settings that could be the cause somehow?

I have had the intention of deprecating the use of django-appconf since it is an extra dependency and can potentially break things and be uncompatible with future Django versions.

Some insight into initialization process is available in the Django documentation

https://docs.djangoproject.com/en/3.1/ref/applications/#initialization-process

**Comment by aleksihakli** [comment]
#640 could potentially fix this, I'm baking an new release version for it so you can test it via PyPI installation.

**Comment by aleksihakli** [comment]
Can you see if 5.6.0 functions correctly?

**Comment by jeremyrich** [comment]
> @jeremyrich Can you see if 5.6.0 functions correctly?

Will check that as soon as possible ! Thank you in advance for Time and support.

**Comment by jeremyrich** [comment]
@aleksihakli It does not seems to fix the issue, still having `AttributeError: 'Settings' object has no attribute 'AXES_LOGGER'`

**Comment by jeremyrich** [comment]
full trace: 
Traceback (most recent call last):
  File "manage.py", line 38, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/usr/local/lib/python3.7/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.7/site-packages/django/apps/registry.py", line 122, in populate
    app_config.ready()
  File "/var/www/jung/badoom/core/apps.py", line 20, in ready
    autodiscover_modules("signals")
  File "/usr/local/lib/python3.7/site-packages/django/utils/module_loading.py", line 47, in autodiscover_modules
    import_module('%s.%s' % (app_config.name, module_to_search))
  File "/usr/local/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.7/site-packages/axes/signals.py", line 15, in <module>
    from axes.handlers.proxy import AxesProxyHandler
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/axes/handlers/proxy.py", line 16, in <module>
    log = getLogger(settings.AXES_LOGGER)
  File "/usr/local/lib/python3.7/site-packages/django/conf/__init__.py", line 80, in __getattr__
    val = getattr(self._wrapped, name)
AttributeError: 'Settings' object has no attribute 'AXES_LOGGER'

**Comment by aleksihakli** [comment]
That's a bit strange. Are you sure that your package caches and old installations etc. are cleared and you are running the 5.6.0 version? The `axes.conf` should be part of that stacktrace as well since the `settings` module is loaded and defaults are set there at the moment. `/var/www/jung/badoom` is showing there so I have a suspicion that this _might_ possibly be linked to some old package state or cache as well and closing out that possibility would be nice. E.g. a clean Docker build-and-run configuration that produces this error is easy to debug because it's easy to share and reproduce.

If your error is up-to-date I think the error is a matter of unconfigured settings at import time. It seems like the Django app loading and import process does not have things configured and the loading of the `axes.handlers.proxy` module is somehow participating in the malfunctioning import chain? The proxy module is being loaded by the `signals` module before the settings are available, maybe? We will have to look into either refactoring some code and getting more familiar with the Django bootstrapping and application loading chain.

The hard thing is that in order to fix this reliably we'd need to be able to reproduce the error and in order to that we would need the "minimal error configuration" which means the minimal clean Django settings that produce this error in a test environment like an example project and settings. If you can test and supply that minimal configuration (e.g. see if just setting `MIGRATION_MODULES` produces this?) then that would help a load. I haven't begun working on that yet.

**Comment by jeremyrich** [comment]
I am quite sure I was in a clean state since I am running it on docker and need to rebuild everything when adding or changing a library. But let me try again to be 💯 sure! -- coming back to you on that Tomorrow. 

I understand that having a minimal error configuration to test would help you a lot. Let me try to Make a clean repo with only Django and axes with this migration process to try to reproduce it.

**Comment by aleksihakli** [comment]
Great, that would be awesome. I'm interested in solving this as well as the problem affects both UX and implies a configuration or cyclic import problem you mentioned in some part of the stack. Clean solution just requires a bit more understanding and research on the subject.

**Comment by jeremyrich** [comment]
@aleksihakli  please find here a minimal django project that reproduce the error : 
https://github.com/jeremyrich/temp-django-axes-error

**Comment by jeremyrich** [comment]
To reproduce just run a simple `python manage.py test` !

**Comment by jeremyrich** [comment]
traceback :
Traceback (most recent call last):
  File "manage.py", line 21, in <module>
    main()
  File "manage.py", line 17, in main
    execute_from_command_line(sys.argv)
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/django/apps/registry.py", line 122, in populate
    app_config.ready()
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/axes/apps.py", line 52, in ready
    from axes import checks, signals  # noqa
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/axes/checks.py", line 8, in <module>
    from axes.backends import AxesBackend
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/axes/backends.py", line 7, in <module>
    from axes.handlers.proxy import AxesProxyHandler
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/axes/handlers/proxy.py", line 16, in <module>
    log = getLogger(settings.AXES_LOGGER)
  File "/home/mymy/Documents/temp-django-axes-error/temp_env/lib/python3.8/site-packages/django/conf/__init__.py", line 80, in __getattr__
    val = getattr(self._wrapped, name)
AttributeError: 'Settings' object has no attribute 'AXES_LOGGER'

**Comment by aleksihakli** [comment]
Great, that makes debugging and backtracing the error a lot easier. I can reproduce the error locally. I'll see if I can identify what's going wrong with the `MIGRATION_MODULES` configured :)

**Comment by jeremyrich** [comment]
Hello @aleksihakli,
did you had chance to find what could be the problem ?

**Comment by jeremyrich** [comment]
Hello @aleksihakli,
did you had chance to find what could be the problem ? Sorry for the reminder, but this would really help us sending axes to production ...

**Comment by aleksihakli** [comment]
Hey @jeremyrich! Haven't had time to take a look into this yet. If you are experiencing this as a blocker please feel free to open a PR with a fix. We welcome all the effort and help people can spare us at Jazzband, being that the community and the projects are mostly free-of-charge and voluntary work based.

**Comment by jeremyrich** [comment]
Hi @aleksihakli ! Well I am not yet an expert in django/python, only a junior developper -- so that's my first participation to a public/community project ! 
I did this PR : https://github.com/jazzband/django-axes/pull/651
It fixed the conflict I had, so please have a look at it. You are allowed to trash me for the review :rofl:

**Comment by aleksihakli** [comment]
That's interesting. However, we need to maintain compatibility in the tests, and hence need the settings to be overridable, so the design might need to be adjusted slighty :)

**Comment by aleksihakli** [comment]
First of all, thanks for the help in locating and reproducing the issue! I tested the current master as well as #652 against the `temp-django-axes-error` project which you kindly supplied, and this bug should be fixed in version 5.7.0 which will be released to PyPI shortly. Would you mind testing if the fix works as intended for you? 

I ran `python manage.py test`  before and after the patch in the test project and the command fails before and passes after applying the patch and setting `MIGRATION_MODULES = {AppConfig.create(entry).label: None for entry in INSTALLED_APPS}` in `settings.py`. Seems to be caused by wrong import orders and using `axes.conf.settings` before they are available in the Django settings system. I refactored the appconfig and logging initialization and dropped `AXES_LOGGER` as it seems to be a bit unnecessary as the Django logging configuration enables setting and configuring `axes` loggers on the top level with [`dictConfig`](https://docs.djangoproject.com/en/dev/topics/logging/#examples) format where `LOGGING["loggers"]["axes"]` package and individual modules can be configured as needed.

**Comment by jeremyrich** [comment]
Thanks a lot for the fix. I will test it ASAP and keep you informed!

I went to far by refactoring all settings ^^ !

**Comment by aleksihakli** [comment]
Refactoring the settings is something that might be nice to do too, as there are better ways for handling them, so I don't think it's too far :) The bug was fixed with a different change, though, so I thought to close the related issues and PRs in the meanwhile. If you would like to see settings structure improved please feel free to reopen the PR!

**Comment by jeremyrich** [comment]
I will try to take Time to continue the settings refactoring yes. You are very kind and took the time to fix my issue so I would be glad to work on it also even if I have still a lot to learn.

**Comment by jeremyrich** [comment]
The fix is working ! :tada: 

I managed to remove those default settings from my project settings : 
 AXES_LOGGER = "axes.watch_login" 
AXES_USERNAME_FORM_FIELD = "username"
AXES_USERNAME_CALLABLE = None
AXES_PROXY_ORDER = "left-most"
AXES_PROXY_COUNT = None
AXES_PROXY_TRUSTED_IPS = None
AXES_ONLY_WHITELIST = False
AXES_WHITELIST_CALLABLE = None

Still axes is asking for :
AXES_LOCK_OUT_BY_USER_OR_IP = False

But that's way better than before ! I am now more confident to push axes in production

**Comment by jeremyrich** [comment]
In fact it's not fixing everything :cry: 

It's working on the temp github I made to reproduce the initial error, but on our project which has different settings files for each env, it's making this : 


  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/var/www/jung/badoom/settings/settings/eu_dev.py", line 1, in <module>
    from .eu_base import *  # noqa
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "/var/www/jung/badoom/settings/settings/eu_base.py", line 8, in <module>
    from .base import *  # noqa
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "/var/www/jung/badoom/settings/settings/base.py", line 866, in <module>
    MIGRATION_MODULES = {AppConfig.create(entry).label: None for entry in INSTALLED_APPS}
  File "/var/www/jung/badoom/settings/settings/base.py", line 866, in <dictcomp>
    MIGRATION_MODULES = {AppConfig.create(entry).label: None for entry in INSTALLED_APPS}
  File "/usr/local/lib/python3.7/site-packages/django/apps/config.py", line 116, in create
    mod = import_module(mod_path)
  File "/usr/local/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/axes/apps.py", line 6, in <module>
    from axes.conf import settings
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 215, in wrapped_find_and_load_unlocked
    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
  File "/usr/local/lib/python3.7/site-packages/ddtrace/internal/import_hooks.py", line 171, in exec_and_call_hooks
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/axes/conf.py", line 6, in <module>
    settings.AXES_ENABLED = getattr(settings, "AXES_ENABLED", True)
  File "/usr/local/lib/python3.7/site-packages/django/conf/__init__.py", line 79, in __getattr__
    self._setup(name)
  File "/usr/local/lib/python3.7/site-packages/django/conf/__init__.py", line 66, in _setup
    self._wrapped = Settings(settings_module)
  File "/usr/local/lib/python3.7/site-packages/django/conf/__init__.py", line 176, in __init__
    raise ImproperlyConfigured("The SECRET_KEY setting must not be empty.")
django.core.exceptions.ImproperlyConfigured: The SECRET_KEY setting must not be empty.

**Comment by aleksihakli** [comment]
@jeremyrich would you mind testing 5.7.1?

I further moved the settings loading into a place where it should work better.

If that doesn't work then I'm suspecting your DataDog library (ddtrace) might be causing some problems with code instrumentation and disabling that and testing things could be a good thing to try 👍

**Comment by jeremyrich** [comment]
Looks perfect !!! 
I managed to removed all the settings :
AXES_USERNAME_FORM_FIELD = "username"
AXES_USERNAME_CALLABLE = None
AXES_PROXY_ORDER = "left-most"
AXES_PROXY_COUNT = None
AXES_PROXY_TRUSTED_IPS = None
AXES_ONLY_WHITELIST = False
AXES_WHITELIST_CALLABLE = None
AXES_LOCK_OUT_BY_USER_OR_IP = False

Thank you for the second fix ! You are really awesome !

**Comment by aleksihakli** [comment]
Good that everything works now! Seems like we might have fixed a potential few bugs along the way as well. Thanks for the help in replicating and testing this, really helps a maintainer when there's a stacktrace and examples available 👍
